fix(kubernetes): clear user token from kube token cache on logout + update cluster rolebindings for user on change of team/user authorization [EE-6298]

prabhat83 opened the pull request.

hookenz commented on api/http/handler/teammemberships/teammembership_delete.go:
```diff
@@ -23,7 +23,7 @@ import (
 // @failure 403 "Permission denied"
 // @failure 404 "TeamMembership not found"
-// @failure 500 "Server error"
+// @failure 500 "Server erzror"
```
erzror Typo?

hookenz commented on api/http/proxy/factory/kubernetes/transport.go:
```diff
@@ -49,7 +49,61 @@ func (transport *baseTransport) proxyKubernetesRequest(request *http.Request) (*
+				endpoint, err := transport.dataStore.Endpoint().Endpoint(portainer.EndpointID(endpointID))
+				if err != nil {
+					log.Error().Err(err).Msgf("failed fetching endpoint %d", endpointID)
+					return
+				}
+
+				restrictDefaultNamespace := endpoint.Kubernetes.Configuration.RestrictDefaultNamespace
+				userIDs := func() []portainer.UserID {
+					userIDs := make([]portainer.UserID, 0)
+					for u := range endpoint.UserAccessPolicies {
+						userIDs = append(userIDs, u)
+					}
+					for t := range endpoint.TeamAccessPolicies {
+						memberships, _ := transport.dataStore.TeamMembership().TeamMembershipsByTeamID(portainer.TeamID(t))
+						for _, membership := range memberships {
+							userIDs = append(userIDs, membership.UserID)
+						}
+					}
+					return userIDs
+				}
+
+				kubecli, err := transport.k8sClientFactory.GetKubeClient(endpoint)
+				if err != nil {
+					log.Error().Err(err).Msgf("failed fetching kube client for endpoint %d", endpointID)
+					return
+				}
+				for _, userID := range userIDs() {
+					memberships, _ := transport.dataStore.TeamMembership().TeamMembershipsByUserID(portainer.UserID(userID))
+					teamIds := make([]int, 0, len(memberships))
+					for _, membership := range memberships {
+						teamIds = append(teamIds, int(membership.TeamID))
+					}
+
+					err := kubecli.SetupUserServiceAccount(int(userID), teamIds, restrictDefaultNamespace)
+					if err != nil {
+						log.Error().Err(err).Msgf("failed setting-up service account for user %d", userID)
+					}
+				}
+			}
+		}()
```
Suggest we put this into a separate function, it's too big for a defer

prabhat83 replied:
I was already doing that before standup. Let me push the change.

hookenz commented on api/http/proxy/factory/kubernetes/transport.go:
```diff
@@ -49,7 +49,61 @@ func (transport *baseTransport) proxyKubernetesRequest(request *http.Request) (*
 	requestPath := apiVersionRe.ReplaceAllString(request.URL.Path, "")
 
+	endpointRe := regexp.MustCompile(`([0-9]+)`)
+	endpointIDMatch := endpointRe.FindAllString(request.RequestURI, -1)
```
Pass 1 as the second argument will mean it'll match a max of 1.  Since you're only wanting the first one and throw away the rest -1 is for all.

hookenz reviewed (changes requested):
A couple of minor things and one typo that definitely needs fixing.

prabhat83 committed:
clear user token from kube token cache on logout

In addtion to that it updates cluster roles on change user/teams authorizations

hookenz commented on api/dataservices/interface.go:
```diff
@@ -103,6 +103,7 @@ type (
 		DeleteEndpoint(ID portainer.EndpointID) error
 		GetNextIdentifier() int
 		BucketName() string
+		EndpointsByTeamID(teamID portainer.TeamID) ([]portainer.Endpoint, error)
```
I recommend you move this function to above `Heartbeat` and below `EndpointIDByEdgeID` .
i.e. to keep related and similar functions together.

hookenz commented on api/http/proxy/factory/kubernetes/transport.go:
```diff
@@ -49,7 +49,19 @@ func (transport *baseTransport) proxyKubernetesRequest(request *http.Request) (*
+	log.Debug().Msgf("proxying request to %s, method %s", request.URL.Path, request.Method)
+
 	switch {
+	case requestPath == "/namespaces/portainer/configmaps/portainer-config" && (request.Method == "PUT" || request.Method == "POST"):
+		defer security.UpdateUserServiceAccountsForEndpoint(portainer.EndpointID(endpointID), transport.dataStore, transport.k8sClientFactory)
+		return transport.executeKubernetesRequest(request)
```
Previously the request has prefix "/namespaces" would call `proxyNamespacedRequest`.  Now it calls `executeKubernetesRequest` ?   It looks more correct that this test should be below the 

`strings.HasPrefix(requestPath, "/namespaces"):` condition.

Where you should call "if requestPath == "/namespaces/portainer/configmaps/portainer-config" && ...

In which case I would move the defer and this if condition inside "proxyNamespacedRequest".   

What do you think?

hookenz commented on api/http/security/authorization.go:
```diff
@@ -160,3 +163,44 @@ func AuthorizedAccess(userID portainer.UserID, memberships []portainer.TeamMembe
+func UpdateUserServiceAccountsForEndpoint(endpointID portainer.EndpointID, dataStore dataservices.DataStore, k8sClientFactory *cli.ClientFactory) {
+	endpoint, err := dataStore.Endpoint().Endpoint(endpointID)
+	if err != nil {
+		log.Error().Err(err).Msgf("failed fetching endpoint %d", endpointID)
```
failed fetching `environment` ...?

hookenz commented on api/http/security/authorization.go:
```diff
@@ -160,3 +163,44 @@ func AuthorizedAccess(userID portainer.UserID, memberships []portainer.TeamMembe
+		return
+	}
+
+	restrictDefaultNamespace := endpoint.Kubernetes.Configuration.RestrictDefaultNamespace
+	userIDs := func() []portainer.UserID {
+		userIDs := make([]portainer.UserID, 0)
+		for u := range endpoint.UserAccessPolicies {
+			userIDs = append(userIDs, u)
+		}
+		for t := range endpoint.TeamAccessPolicies {
+			memberships, _ := dataStore.TeamMembership().TeamMembershipsByTeamID(portainer.TeamID(t))
+			for _, membership := range memberships {
+				userIDs = append(userIDs, membership.UserID)
+			}
+		}
+		return userIDs
+	}
```
This should be in it's own function.

hookenz commented on api/http/security/authorization.go:
```diff
@@ -160,3 +163,44 @@ func AuthorizedAccess(userID portainer.UserID, memberships []portainer.TeamMembe
 	return false
 }
+
+func UpdateUserServiceAccountsForEndpoint(endpointID portainer.EndpointID, dataStore dataservices.DataStore, k8sClientFactory *cli.ClientFactory) {
```
Consider using a receiver. i.e. can then be called like `endpoint.UpdateUserServiceAccounts(dataStore, k8sClientFactory) `

```
func (endpointID portainer.EndpointID)UpdateServiceAccounts(dataStore ...
```

hookenz reviewed (approved):
Approved, but I added some suggested changes.   One of which could be a bug

prabhat83 committed:
address review comments

prabhat83 committed:
address review comments

prabhat83 committed:
address review comments

prabhat83 committed:
fixed
